components/batch-system/benches/batch-system.rs

```diff
@@ -0,0 +1,130 @@
+    const ID_LIMIT: u64 = 32;
+    const MESSAGE_LIMIT: usize = 256;
+    for id in 0..ID_LIMIT {
+        let (normal_tx, normal_fsm) = Runner::new(100000);
+        let normal_box = BasicMailbox::new(normal_tx, normal_fsm);
+        router.register(id, normal_box);
+    }
+
+    let (tx, rx) = std::sync::mpsc::channel();
+    c.bench_function("spawn_many", |b| {
+        b.iter(|| {
+            for id in 0..ID_LIMIT {
+                for i in 0..MESSAGE_LIMIT {
+                    router.send(id, Message::Loop(i)).unwrap();
+                }
+                router.send(id, end_hook(&tx)).unwrap();
+            }
+            for _ in 0..ID_LIMIT {
+                rx.recv().unwrap();
+            }
+        })
+    });
+    system.shutdown();
+}
+
+fn bench_imbalance(c: &mut Criterion) {
+    let (control_tx, control_fsm) = Runner::new(100000);
+    let (router, mut system) = batch_system::create_system(2, 2, control_tx, control_fsm);
+    system.spawn("test".to_owned(), Builder::new());
+    const ID_LIMIT: u64 = 10;
+    const MESSAGE_LIMIT: usize = 512;
+    for id in 0..ID_LIMIT {
+        let (normal_tx, normal_fsm) = Runner::new(100000);
+        let normal_box = BasicMailbox::new(normal_tx, normal_fsm);
+        router.register(id, normal_box);
+    }
+
+    let (tx, rx) = std::sync::mpsc::channel();
+    c.bench_function("imbalance", |b| {
+        b.iter(|| {
+            for i in 0..MESSAGE_LIMIT {
+                for id in 0..2 {
+                    router.send(id, Message::Loop(i)).unwrap();
+                }
+            }
+            for id in 0..2 {
+                router.send(id, end_hook(&tx)).unwrap();
+            }
+            for _ in 0..2 {
+                rx.recv().unwrap();
+            }
+        })
+    });
+    system.shutdown();
+}
+
+fn bench_fairness(c: &mut Criterion) {
+    let (control_tx, control_fsm) = Runner::new(100000);
+    let (router, mut system) = batch_system::create_system(2, 2, control_tx, control_fsm);
+    system.spawn("test".to_owned(), Builder::new());
+    for id in 0..10 {
+        let (normal_tx, normal_fsm) = Runner::new(100000);
+        let normal_box = BasicMailbox::new(normal_tx, normal_fsm);
+        router.register(id, normal_box);
+    }
+
+    let (tx, _rx) = std::sync::mpsc::channel();
+    let running = Arc::new(AtomicBool::new(true));
+    let router1 = router.clone();
+    let running1 = running.clone();
+    let handle = std::thread::spawn(move || {
+        while running1.load(Ordering::SeqCst) {
+            for id in 0..4 {
+                let _ = router1.send(id, Message::Loop(16));
+            }
+        }
+        tx.send(()).unwrap();
+    });
+
+    let (tx2, rx2) = std::sync::mpsc::channel();
+    c.bench_function("fairness", |b| {
+        b.iter(|| {
+            for _ in 0..10 {
+                for id in 4..6 {
+                    router.send(id, Message::Loop(10)).unwrap();
+                }
+            }
+            for id in 4..6 {
```

Comment:
I'm a bit confused by these ids - are these two bench functions meant to be run at the same time? Do they interact? Is it deliberate or coincidence that the sets of ids are disjoint? I think that if there is some relation, then it needs to be documented